### PR TITLE
add test locking removed when amount is zero

### LINKED ISSUE #14231: balances pallet: test locks of zero
Add a test checking the lock is properly removed when `set_lock` is called with amount zero 
 https://github.com/paritytech/substrate/pull/14144/files#diff-0848af5dac3286b1cf09718d38571925e65582729dedacceef836bb2278b612dR857

---

              someone needs to write some tests for this

_Originally posted by @xlc in https://github.com/paritytech/substrate/issues/14144#issuecomment-1560243663_

### PR DESCRIPTION
**Pull request by jpserrat**
Closes #14231 

Hey @liamaharon, I've added the test when the amount is zero. Let me know if this is what you have in mind.

### CODE REVIEW

File: frame/balances/src/tests/currency_tests.rs [hunk 1]
```diff
@@ -33,6 +33,18 @@ const ID_2: LockIdentifier = *b"2       ";
 pub const CALL: &<Test as frame_system::Config>::RuntimeCall =
 	&RuntimeCall::Balances(crate::Call::transfer_allow_death { dest: 0, value: 0 });
 
+#[test]
+fn locking_should_be_removed_when_amount_is_zero() {
```
Comment (liamaharon) [inline frame/balances/src/tests/currency_tests.rs:37]:
```suggestion
fn setting_zero_amount_removes_lock() {
```

**Review by liamaharon — COMMENTED** [review]
Looks good! While you're at it, could you please also check setting reasons to empty (`WithdrawReasons::empty()`) has the same behavior? See https://github.com/paritytech/substrate/pull/14144/files#diff-0848af5dac3286b1cf09718d38571925e65582729dedacceef836bb2278b612dR856-R859.

**Comment by Jpserrat** [comment]
Hey @liamaharon, thanks for the quick response! Just did the requested changes.

File: frame/balances/src/tests/currency_tests.rs [hunk 2]
```diff
@@ -33,6 +33,30 @@ const ID_2: LockIdentifier = *b"2       ";
...
+}
+
+#[test]
+fn set_lock_with_withdraw_reasons_empty_removes_lock() {
+	ExtBuilder::default()
+	    .existential_deposit(1)
+		.monied(true)
+		.build_and_execute_with(|| {
+			Balances::set_lock(ID_1, &1, u64::MAX, WithdrawReasons::all());
+			Balances::set_lock(ID_1, &1, 2, WithdrawReasons::empty());
```
Comment (liamaharon) [inline frame/balances/src/tests/currency_tests.rs:55]:
```suggestion
			Balances::set_lock(ID_1, &1, u64::MAX, WithdrawReasons::empty());
```

**Review by liamaharon — APPROVED** [review]
Tabs are preferred for indentation, before merging please fix by running `cargo +nightly fmt -p pallet-balances`

**Review by ggwpez — APPROVED** [review]

**Comment by ggwpez** [comment]
bot merge
